### ISSUE TITLE
c++ gradient is not implemented for concat/stack

### ISSUE THREAD

**Issue by tcmxx** [tensorflow/tensorflow#19944]
Feature Request:
I noticed that c++ gradient for concat or stack is not implemented while i am using tensorflow sharp. I think those are quite important operations.  Could you please somehow consider implement those? Or is there any other replacement?

Thank you!

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by tcmxx** [comment]
System information
Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No
OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Windows 10
TensorFlow installed from (source or binary):
pip and built from coursecode
TensorFlow version (use command below):
1.8.0
Python version:
3.6
Bazel version (if compiling from source):
0.14.1
GCC/Compiler version (if compiling from source):
VS??
CUDA/cuDNN version:
N/A
GPU model and memory:
Nvdia 960m
Exact command to reproduce:
NA

**Comment by reedwm** [comment]
Are you sure the concat grad is not implemented? I think it's implemented [here](https://github.com/tensorflow/tensorflow/blob/8212404a47e17a0ad1822e520c990be1cd712e91/tensorflow/core/ops/array_grad.cc#L224).

**Comment by reedwm** [comment]
/CC @suharshs, any idea why gradients wouldn't be included?

**Comment by suharshs** [comment]
Some operations aren't implemented yet. Pull requests are welcome and instructions for adding new gradients are available here: https://www.tensorflow.org/code/tensorflow/cc/gradients/README.md

**Comment by tcmxx** [comment]
@suharshs Does your team have any plan implementing the gradient for concat or stack in the near future? I would love to implement it myself but the code seems not very intuitive...

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
@tcmxx Currently this has not been prioritized.

If you want to give it a try, you can look at some previous open source contributions: https://github.com/tensorflow/tensorflow/pull/17592/files to gain some inspiration :)

**Comment by tcmxx** [comment]
@suharshs Ok. I will give a try soon.

**Comment by tcmxx** [comment]
Just for anyone who is interested, I made a very simplified version to be added to array_grad.cc based on python implementation....sanity checks/optimizations are removed, not working for all cases....


```

std::vector<Output> ExtractInputShapes(const Scope& scope, const std::vector<Output>& inputs){
  //Extract the shapes of a set of input tensors.
  bool fully_known = true;
  std::vector<Output> sizes;
  for(int i = 0; i < inputs.size();++i){
    auto input_shape = Shape(scope, inputs[i]);
    //removed because I dont konw how to check the type in c++...Assume that fully_know is always true..
    //if(dynamic_cast<Const*>((&input_shape.op())) == nullptr || dynamic_cast<Tensor*>((&input_shape)) == nullptr){
    /*if(dynamic_cast<Tensor*>((&input_shape)) == nullptr){
      fully_known = false;
      break;
    }*/
    sizes.push_back(input_shape);
  }

  if(fully_known){
    return sizes;
  }else{
    auto result = ShapeN(scope, InputList(inputs)).output;
    return result;
  }

}

Status ConcatGradHelper(const Scope& scope, const Operation& op,
                 const std::vector<Output>& grad_inputs,
                 std::vector<Output>* grad_outputs,
               const int start_value_index, const int end_value_index, const int dim_index) {
  // Degenerate concatenation, just return grad.
  if(op.num_inputs() == 2){
    if(end_value_index > dim_index){
      grad_outputs->push_back(NoGradient());
    }
    for (int i = 0; i < grad_inputs.size();++i){
      grad_outputs->push_back(grad_inputs[i]);
    }
    if(end_value_index <= dim_index){
      grad_outputs->push_back(NoGradient());
    }
    return scope.status();
  }

  auto concat_dim = op.input(dim_index);
  std::vector<Output> input_values;
  for(int i = 0; i < end_value_index - start_value_index;++i){
    input_values.push_back(op.input(i+start_value_index));
  }
  std::vector<Output> out_grads;

  //No support for sparse or eagerly executing
  auto non_neg_concat_dim = FloorMod(scope, concat_dim, Rank(scope, input_values[0]));
  auto sizes = ExtractInputShapes(scope, input_values);


  //temperary use this implementation only...the python implementation uese two different ones for different input size for optimization....
  auto stackedSized= Stack(scope,sizes,Stack::Axis(1));
  std::vector<Output> sliceBegin = {non_neg_concat_dim.z, Const(scope, 0)};
  auto sliceBeginStacked = Stack(scope,InputList(sliceBegin),Stack::Axis(0));
  auto slicedSizes = Slice(scope,stackedSized, sliceBeginStacked,{1, -1});
  auto squeezedSizes = Squeeze(scope,slicedSizes);
  out_grads = SplitV(scope,grad_inputs[0],squeezedSizes,  non_neg_concat_dim.z, sizes.size()).output;

  if(end_value_index > dim_index){
    grad_outputs->push_back(NoGradient());
  }
  for (int i = 0; i < out_grads.size();++i){
    grad_outputs->push_back(out_grads[i]);
  }
  if(end_value_index <= dim_index){
    grad_outputs->push_back(NoGradient());
  }
  return scope.status();
}

Status ConcatGrad(const Scope& scope, const Operation& op,
                    const std::vector<Output>& grad_inputs,
                    std::vector<Output>* grad_outputs){
  return ConcatGradHelper(scope, op,grad_inputs,grad_outputs,1,op.num_inputs(),0);
}
REGISTER_GRADIENT_OP("Concat", ConcatGrad);

Status ConcatGradV2(const Scope& scope, const Operation& op,
                    const std::vector<Output>& grad_inputs,
                    std::vector<Output>* grad_outputs){
  return ConcatGradHelper(scope, op,grad_inputs,grad_outputs,0,op.num_inputs()-1,op.num_inputs()-1);
}
REGISTER_GRADIENT_OP("ConcatV2", ConcatGradV2);
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tcmxx** [comment]
Hi @suharshs This implementation is not as good as the python implementation because it not optimized, and I am not sure if it is usable under all circumstances, so I was not planning to create a pull request, unless you think it is ok.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by gordoncaleb** [comment]
Hi @tcmxx this would be very useful for me as well. Any chance you'll raise a PR or do you mind if I try and pick it up?

**Comment by tcmxx** [comment]
@gordoncaleb Nice to hear it is helpful. You can try and pick it up as you want, just keep me updated~

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by hdu-hh** [comment]
**ConcatGradient** has [already been implemented](https://github.com/tensorflow/tensorflow/commit/084075e4e967677fc0be9fe921a0157126e69617) in 2016 and has been in many Tensorflow releases from before version 1.0 and is there today.

The error **No gradient defined for op: Concat** seems to come from the Lookup() method of a **GradOpRegistry** singleton. The gradient functions managed there are registered via a `REGISTER_GRADIENT_OP` macro.

The existing **ConcatGradient** function is registered via `REGISTER_OP_GRADIENT` instead, which feeds an **OpGradFactor** singleton.

Now some questions:
- how are the different gradient factories and registries supposed to interact?
- can they be merged?
- is the existing concat gradient just registered wrongly (i.e. should it use the `REGISTER_GRADIENT_OP` instead of the `REGISTER_OP_GRADIENT` macro)? @keveman

**Comment by mohantym** [comment]
Hi @tcmxx ! Is this issue still valid in the 2.x version . It seems to be resolved from the above comment though. Thank you!

**Comment by hdu-hh** [comment]
This issue is still valid (checked with the official libtensorflow-cpu 2.9.0 for x86_64). It reports `failed to add operation "Gradients": No gradient defined for op: Concat.`
This is sorely missing because many upstream models use concat and without gradient-support in the libtensorflow-cpu library they cannot be trained directly.

**Comment by mohantym** [comment]
@hdu-hh ! Thanks for the update .

**Comment by hdu-hh** [comment]
Good news! With [Commit bf3d89b1e8](https://github.com/tensorflow/tensorflow/commit/bf3d89b1e8bc5e00271c2fbd57a8b3ff55d48462) more Gradient Operations where added to Tensorflow C++. Concat is still not supported yet, but ConcatV2 is. If rewriting the graph is an option then replacing the Concat op with ConcatV2 (after adapting its inputs) solves the problem.

**Comment by tilakrayal** [comment]
@tcmxx,
Could you please take a look at the [comment](https://github.com/tensorflow/tensorflow/issues/19944#issuecomment-1196470579) from the user and  kindly update to the latest stable version 2.10 and let us know if you are facing the same issue. Thank you!
